### PR TITLE
[CW-22233] feat(kas): support different script type

### PR DESCRIPTION
**Pull request by gundomlegend**
**PR Summary by Typo**
------------

 **Summary**

This pull request includes multiple changes: updating package versions, introducing new types and functions, refactoring methods, and adding tests for various address versions and script types.

**Key Points**

1. Package versions updated in `package-lock.json` and `package.json`.
2. New `scriptType` property introduced for input, output, change, transaction utxo, and script public key types.
3. `getAddressByPublicKeyOrScriptHash` function modified to return a `Payment` object and `getPubkeyOrScriptHash` and `pubkeyOrScriptHashToPayment` functions introduced.
4. `getAddressAndOutScript` methods refactored to return a `Payment` object.
5. Tests added for handling different address versions and script types. 

 <h6>To turn off PR summary, please visit <a href="https://app.typoapp.io/settings/dev-analytics/notification?tab=codeHealth">Notification settings</a>.</h6>

### CODE REVIEW

**Comment by cbx-tech** [comment]
Task linked: [CW-22233 調整 sdk scriptPublicKey、address、getArgument 相關邏輯](https://app.clickup.com/t/25577573/CW-22233)

File: packages/coin-kas/src/index.ts [hunk 1]
```diff
@@ -16,12 +21,13 @@ export default class KAS extends COIN.ECDSACoin implements COIN.Coin {
...
     appPrivateKey: string,
     appId: string,
-    _scriptType: ScriptType,
+    scriptType: ScriptType,
     addressIndex: number,
     purpose?: number
   ): Promise<string> {
     const publicKey = await this.getPublicKey(transport, appPrivateKey, appId, addressIndex, purpose);
-    return getAddressByPublicKey(publicKey);
+    const { pubkeyOrScriptHash, addressVersion } = getPubkeyOrScriptHash(scriptType, publicKey);
```
Comment (gundomlegend) [inline packages/coin-kas/src/index.ts:29]: 有多過 `getPubkeyOrScriptHash` 判斷 + 產生 `publicKeyOrScriptHash`，若遇到不支援的 `scriptType` 會報錯

File: packages/coin-kas/src/index.ts [hunk 2]
```diff
@@ -31,30 +37,34 @@ export default class KAS extends COIN.ECDSACoin implements COIN.Coin {
...
     for (const input of inputs) {
       const pubkey = await this.getPublicKey(transport, appPrivateKey, appId, input.addressIndex, input.purposeIndex);
       input.pubkeyBuf = Buffer.from(pubkey, 'hex');
+      input.scriptType = scriptType;
     }
     if (change) {
       const pubkey = await this.getPublicKey(transport, appPrivateKey, appId, change.addressIndex, change.purposeIndex);
       change.pubkeyBuf = Buffer.from(pubkey, 'hex');
+      change.scriptType = scriptType;
     }
```
Comment (gundomlegend) [inline packages/coin-kas/src/index.ts:68]: input、change 的 scriptType 來源一致

File: packages/coin-kas/src/transaction/index.ts [hunk 3]
```diff
@@ -32,23 +36,29 @@ export class Transaction {
...
-      const pubKey = input.pubkeyBuf?.toString('hex') as string;
       const inputPreValueBN = new BigNumber(input.preValue);
+      const pubKey = input.pubkeyBuf?.toString('hex') as string;
+      const inputScriptType = input.scriptType as ScriptType;
+      const { pubkeyOrScriptHash, addressVersion } = getPubkeyOrScriptHash(inputScriptType, pubKey);
       this.utxos.push({
-        pkScript: pubkeyToPayment(pubKey).outScript,
+        version: 0,
+        pkScript: pubkeyOrScriptHashToPayment(pubkeyOrScriptHash, addressVersion).outScript,
+        scriptType: inputScriptType,
```
Comment (gundomlegend) [inline packages/coin-kas/src/transaction/index.ts:46]: 此處的 version 是 scriptVersion，非 addressVersion，此外有多傳 scriptType 進去

File: packages/coin-kas/src/transaction/index.ts [hunk 4]
```diff
@@ -32,23 +36,29 @@ export class Transaction {
...
     const output = txData.output;
     const outputValueBN = new BigNumber(output.value);
+    const { scriptType: outputScriptType, outScript, outPubkeyOrHash } = addressToOutScript(output.address);
     this.outputs.push({
       scriptPublicKey: {
         version: 0,
-        scriptPublicKey: toHex(addressToOutScript(output.address).outScript),
+        scriptType: outputScriptType,
+        scriptPublicKey: toHex(outScript),
+        publicKeyOrScriptHash: toHex(outPubkeyOrHash),
```
Comment (gundomlegend) [inline packages/coin-kas/src/transaction/index.ts:61]: scriptPublicKey 只有用在跑測試驗證 hash，卡片實際上會用 publicKeyOrScriptHash 在裡面轉換成 scriptPublicKey

File: packages/coin-kas/src/utils/__test__/address.test.ts [hunk 5]
```diff
@@ -1,117 +1,107 @@
...
-    } catch (error) {
-      expect(error).toMatchInlineSnapshot(
-        `[Error: error function: decodeAddress, message: Wrong public key from address: kaspa:qyp2d6tctf8wkahskz4hv3xpe3llknrc3jl66798elnjvr2gewzrdggx6fnfh3y]`
-      );
-    }
+  it('Test decodeAddress throws error if address with unknown version', () => {
+    const address = 'kaspa:q2cm3y2xe65ne797cmar6ntecfjcdtqf585whum65kgyv20k8jzhmtq5q5qe6c9';
+    expect(() => decodeAddress(address)).toThrowErrorMatchingInlineSnapshot(
+      `"error function: decodeAddress, message: Unsupported version: 2 with address:kaspa:q2cm3y2xe65ne797cmar6ntecfjcdtqf585whum65kgyv20k8jzhmtq5q5qe6c9"`
+    );
```
Comment (gundomlegend) [inline packages/coin-kas/src/utils/__test__/address.test.ts:105]: 新增其他格式地址和改顯示字串而非 Buffer

File: packages/coin-kas/src/utils/__test__/validate.test.ts [hunk 6]
```diff
@@ -143,7 +143,7 @@ describe('Test validate.ts', () => {
     it('Test transaction with too large mass', async () => {
       const transaction = Transaction.fromTxData(testTxDataWithLargeInputs);
       expect(() => validateTransaction(transaction)).toThrowErrorMatchingInlineSnapshot(
-        `"error function: validateMass, message: validate: current mass: 101038, exceeds the maximum allowed mass(100000)"`
+        `"error function: validateMass, message: validate: current mass: 940038, exceeds the maximum allowed mass(100000)"`
```
Comment (gundomlegend) [inline packages/coin-kas/src/utils/__test__/validate.test.ts:146]: sigOpCount 的邏輯有算錯導致這邊顯示錯，原先不管多少 input 都是 1

**Review by hankliu418 — APPROVED** [review]

**Review by tp26610 — APPROVED** [review]

**Pull request merged**